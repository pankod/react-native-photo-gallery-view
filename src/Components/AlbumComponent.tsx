// Global Imports
import React, { PureComponent } from 'react';
import { Text, View, Image, TouchableOpacity, Dimensions, FlatList, Animated } from 'react-native';

// Local Imports
import { BlurImage } from '@Components';
import { AlbumStyle } from "@Styles";
import { IAlbumProps, IMediaItem, IGalleryProps, IGalleryState } from "@Interfaces";
import Common from '@Provider';

export class AlbumComponent extends PureComponent<IAlbumProps, {}> {

	static contextType = Common;
	public animatedY = new Animated.Value(0);

	public render(): JSX.Element {
		return (
			<Common.Consumer>
				{
					(context: IGalleryProps & IGalleryState) => (
						<View style={AlbumStyle.container}>
							<FlatList
								style={AlbumStyle.container}
								data={context.mediaList}
								numColumns={context.gridSize}
								renderItem={({ item, index }) => this.renderItem(item, index)}
								keyExtractor={(item, index) => index.toString()}
								extraData={[context.selected, context.dynamicSize]}
							/>
						</View>
					)
				}
			</Common.Consumer>
		)
	}

	private renderItem(item: IMediaItem, index: number): JSX.Element {
		const {
			gridSize,
			onSelection,
			displaySelectionButtons,
			showImageModal,
			stickyFooter,
			renderCustomState,
			dynamicSize
		} = this.context;
		if (displaySelectionButtons && stickyFooter) {
			return (
				<TouchableOpacity
					onPress={() => onSelection(item, index)} key={index}
					style={
						[
							this.checkedCtrl(item) && AlbumStyle.checkedBorder,
							{ position: "relative", width: dynamicSize.width, height: dynamicSize.height, padding: 3 }
						]
					}>

					{this.isChecked(item)}
					<BlurImage resizeMode={"cover"} style={{ flex: 1 }} key={index} source={{ uri: item.thumb }} />
				</TouchableOpacity>
			)
		} else {
			return (
				<TouchableOpacity
					onPress={() => showImageModal(item, index)} key={index}
					style={
						{ width: dynamicSize.width, height: dynamicSize.height, padding: 3 }
					}
				>
					{renderCustomState(item, index)}
					<BlurImage resizeMode={"cover"} style={{ flex: 1 }} key={index} source={{ uri: item.thumb }} />
				</TouchableOpacity >
			)
		}
	}

	private checkedCtrl(item: IMediaItem): boolean {
		const { selected } = this.context;

		return selected.indexOf(item.id) > -1
	}

	private isChecked(item: IMediaItem): JSX.Element {
		const { customCheckedView } = this.context;
		if (this.checkedCtrl(item)) {

			if (customCheckedView) {
				return customCheckedView()
			}

			return (
				<View style={AlbumStyle.checkedContainer}>
					<Image
						style={AlbumStyle.checkedImage}
						source={require("../Assets/Images/checked.png")}
					/>
				</View>
			)
		}

		return null;
	}
}
