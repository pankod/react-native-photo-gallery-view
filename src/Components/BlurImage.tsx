// Global Imports
import React, { PureComponent } from 'react'
import { Animated } from 'react-native'

// Local Imports
import { IBlurImageProps } from '@Interfaces';
import { BlurImageStyle } from '@Styles';

export class BlurImage extends PureComponent<IBlurImageProps> {

	public imageAnimated = new Animated.Value(0);

	private onImageLoad(): void {
		Animated.timing(this.imageAnimated, {
			toValue: 1,
			duration: 500,
			useNativeDriver: true
		}).start();
	}

	render() {
		return (
			<React.Fragment>
				<Animated.Image
					{...this.props}
					style={[BlurImageStyle.image, { opacity: this.imageAnimated }]}
					onLoad={this.onImageLoad.bind(this)}
					blurRadius={1}
				/>
			</React.Fragment>
		)
	}
}
