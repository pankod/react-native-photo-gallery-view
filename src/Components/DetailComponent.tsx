// Global Imports
import React, { PureComponent } from 'react'

// Local Imports
import Common from '@Provider';
import { IGalleryProps, IGalleryState } from '@Interfaces';
import { BlurImage } from '@Components';

export class DetailComponent extends PureComponent {

	static contextType = Common;

	public render(): JSX.Element {
		return (
			<Common.Consumer>
				{
					(context: IGalleryProps & IGalleryState) => (
						<BlurImage
							resizeMethod={"resize"}
							resizeMode={"contain"}
							source={{ uri: context.showingImage && context.showingImage.photo }}
							thumbnail={{ uri: context.showingImage && context.showingImage.thumb }}
						/>
					)
				}
			</Common.Consumer>
		)
	}
}
