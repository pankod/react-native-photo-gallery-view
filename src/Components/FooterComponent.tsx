// Global Imports
import React, { PureComponent } from 'react';
import { View } from 'react-native';

// Local Imports
import { FooterStyle } from "@Styles";
import { IFooterProps, IGalleryProps, IGalleryState } from "@Interfaces";
import Common from '@Provider';
import { Const } from '@Constants';

export class FooterComponent extends PureComponent<IFooterProps, {}> {

	static contextType = Common;

	public render(): JSX.Element {
		return (
			<Common.Consumer>
				{
					(context: IGalleryProps & IGalleryState) => (
						<View style={[FooterStyle.container, context.customFooterStyle]}>
							{context.renderStickyFooter && this.renderStickyFooter()}
						</View>
					)
				}
			</Common.Consumer>
		)
	}

	private renderStickyFooter(): JSX.Element {
		return (
			this.context.renderStickyFooter(Const.HEIGHT)
		)
	}
}
