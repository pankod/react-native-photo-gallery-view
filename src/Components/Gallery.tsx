import React, { Component } from "react"
import { Text, View } from "react-native"

export class Gallery extends Component {
	render() {
		return (
			<View>
				<Text> textInComponent </Text>
			</View>
		)
	}
}
