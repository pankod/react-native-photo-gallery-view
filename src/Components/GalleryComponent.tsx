// Global Imports
import React, { Component } from "react"
import { View, Dimensions, BackHandler, AppState } from "react-native"

// Local Imports
import { TopBarComponent, AlbumComponent, FooterComponent, DetailComponent, PreviewModal } from "@Components";
import { IGalleryProps, IGalleryState, IMediaItem } from "@Interfaces";
import { GalleryStyle } from "@Styles";
import Common from '@Provider';

const { width, height } = Dimensions.get('window');

export class GalleryComponent extends Component<IGalleryProps, IGalleryState> {

	constructor(props: IGalleryProps) {
		super(props);
		this.state = {
			title: `${props.mediaList.length} Photos`,
			detailTitle: null,
			showingImage: {} as IMediaItem,
			isModalOpen: false,
			showImageModal: this.showImageModal.bind(this),
			showPreview: this.showPreview.bind(this),
			hidePreview: this.hidePreview.bind(this),
			onBackRequest: this.onBackRequest.bind(this),
			onSelection: this.onSelection.bind(this),
			changeImage: this.changeImage.bind(this),
			imageIndex: 0,
			selected: [],
			orientation: 'portrait',
			dynamicSize: {
				width: width / props.gridSize,
				height: width / props.gridSize,
			},
			previewIsOpen: false,
			imagePreview: {} as IMediaItem
		}
		// this.getOrientation = this.getOrientation.bind(this);
		this.backKeyHandler = this.backKeyHandler.bind(this)
		this.unTouch = this.unTouch.bind(this);
	}

	static defaultProps = {
		gridSize: 3,
		stickyFooter: true,
		displaySelectionButtons: false,
		detailImageResizeMode: "contain",
		detailImageResizeMethod: "resize",
		thumbImageResizeMode: "cover",
		thumbImageResizeMethod: "resize"
	}

	public render(): JSX.Element {
		const {
			style
		} = this.props;
		const { isModalOpen, previewIsOpen } = this.state;
		return (
			<Common.Provider value={{ ...this.state, ...this.props }}>
				<View
					onTouchEnd={() => this.unTouch()}
					// onTouchEndCapture={() => console.log("end capture")}
					ref="rootView"
					style={[GalleryStyle.container, style]}
				>
					<TopBarComponent />
					{!isModalOpen && <AlbumComponent />}
					{isModalOpen && <DetailComponent />}
					<FooterComponent />
					{previewIsOpen && <PreviewModal />}
				</View>
			</Common.Provider>
		)
	}

	public showImageModal(media: IMediaItem, index: number): void {
		const { mediaList } = this.props;
		this.setState({
			imageIndex: index,
			detailTitle: `${index + 1} of ${mediaList.length}`,
			showingImage: media,
			isModalOpen: true
		});
	}

	public changeImage(index: number): void {
		const { mediaList } = this.props;
		this.setState({
			imageIndex: index,
			detailTitle: `${index + 1} of ${mediaList.length}`,
			showingImage: mediaList[index],
		});
	}

	public componentWillReceiveProps(nextProps): void {
		if (nextProps.displaySelectionButtons !== this.props.displaySelectionButtons) {
			this.setState({
				selected: []
			})
		}
	}

	public componentDidMount(): void {
		BackHandler.addEventListener("hardwareBackPress", this.backKeyHandler);
	}

	public componentWillUnmount(): void {
		BackHandler.removeEventListener("hardwareBackPress", this.backKeyHandler);
	}

	public backKeyHandler(): boolean {
		this.onBackRequest();
		return true;
	}

	public onBackRequest(): void {
		const { onBack } = this.props;
		const { isModalOpen } = this.state;
		if (onBack && !isModalOpen) {
			onBack();
		}

		if (isModalOpen) {
			this.clearModal();
		}
	}

	public showPreview(item: IMediaItem, index: number): void {
		this.setState({
			imagePreview: item,
			previewIsOpen: true
		});
	}

	public hidePreview(): void {
		this.setState({
			previewIsOpen: false,
			imagePreview: {} as IMediaItem
		});
	}

	public unTouch(): void {
		if (this.state.previewIsOpen) {
			this.hidePreview();
		}
	}

	public clearModal(): void {
		this.setState({
			imageIndex: 0,
			detailTitle: null,
			showingImage: null,
			isModalOpen: false,
			selected: []
		});
	}

	public onSelection(item: IMediaItem, index: number): void {
		const { onSelectionChanged } = this.props;
		const { selected } = this.state;

		if (onSelectionChanged) {
			onSelectionChanged(item, index);
		}

		if (selected.indexOf(item.id) === -1) {
			selected.push(item.id);
		} else {
			selected.splice(selected.indexOf(item.id), 1);
		}

		this.setState({
			selected: [...selected]
		});
	}
}
