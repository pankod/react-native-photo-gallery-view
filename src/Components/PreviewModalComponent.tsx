// Global Imports
import React, { PureComponent } from 'react'
import { Text, View, Modal, Image, Animated, ImageBackground, Platform } from 'react-native'

// Local Imports
import Common from '@Provider';
import { IGalleryProps, IGalleryState } from '@Interfaces';
import { PreviewModalStyle } from '@Styles';

export class PreviewModal extends PureComponent {

	static contextType = Common;

	public scale = new Animated.Value(0);
	public opacity = new Animated.Value(0);

	public componentDidMount(): void {
		Animated.parallel([
			Animated.spring(this.scale, {
				toValue: 1,
				bounciness: 10,
				useNativeDriver: true
			}),
			Animated.timing(this.opacity, {
				toValue: 1,
				duration: 500,
				useNativeDriver: true
			})
		]).start();
	}

	public render(): JSX.Element {
		return (
			<Common.Consumer>
				{(context: IGalleryProps & IGalleryState) => (
					<Modal
						animationType={"none"}
						visible={context.previewIsOpen}
						transparent={true}
					>
						<ImageBackground style={{
							flex: 1,
						}}
							blurRadius={Platform.OS == 'ios' ? 20 : 15}
							// source={{ uri: context.imagePreview.thumb }}
							source={require("../Assets/Images/radius-bg.png")}
							resizeMode={"cover"}
						>
							<Animated.View
								style={[
									PreviewModalStyle.container,
									{
										transform: [
											{
												scale: this.scale.interpolate({
													inputRange: [0, 1],
													outputRange: [0, 1]
												})
											}
										],
										opacity: this.opacity
									}
								]}>
								<View style={PreviewModalStyle.imagePreview}>
									{
										context.customPreviewComponent ?
											context.customPreviewComponent(context.imagePreview) :
											(
												<Image
													style={{ width: "100%", flex: 1 }}
													source={{ uri: context.imagePreview.photo }}
													resizeMode={"contain"}
												/>
											)
									}
								</View>
							</Animated.View>
						</ImageBackground>
					</Modal>
				)}
			</Common.Consumer>
		)
	}
}
