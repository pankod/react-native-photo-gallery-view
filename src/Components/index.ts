export { GalleryComponent } from '@Components/GalleryComponent';
export { TopBarComponent } from '@Components/TopBarComponent';
export { FooterComponent } from '@Components/FooterComponent';
export { AlbumComponent } from '@Components/AlbumComponent';
export { BlurImage } from '@Components/BlurImage';
export { DetailComponent } from '@Components/DetailComponent';
export { PreviewModalComponent } from '@Components/PreviewModalComponent';
