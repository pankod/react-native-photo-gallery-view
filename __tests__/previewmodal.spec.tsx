
import * as React from 'react';

import renderer from "react-test-renderer";
import { ShallowWrapper, shallow, mount } from "enzyme";

import { PreviewModal } from "../src/Components/PreviewModalComponent";
import Common from "../src/Provider";

describe("PreviewModal Component", () => {
	let wrapper: ShallowWrapper;
	let mounting;
	const props = {
		imagePreview: {
			photo: {}
		},
		renderPreview: jest.fn()
	}
	const state = {}
	let component = (
		<Common.Provider value={{ items: [], columns: 3, ...props, ...state }}>
			<PreviewModal {...props} />
		</Common.Provider>
	);

	beforeEach(() => {
		wrapper = shallow(component, { context: { ...state, ...props } })
		mounting = mount(component, { context: { ...state, ...props } });
	});

	test('should PreviewModal Component is render correctly', () => {
		const render = renderer.create(component).toJSON();

		expect(render).toBeDefined();
		expect(render).toMatchSnapshot();
	});

	test('should hidethumb method', () => {
		const spy = spyOn(PreviewModal.prototype, "hideThumb").and.callThrough();
		const instance = mounting.instance() as any;

		instance.hideThumb();
		expect(spy).toHaveBeenCalled();
	});

	test('should render defaultView', () => {
		const spy = spyOn(PreviewModal.prototype, "defaultView").and.callThrough();
		const instance = mounting.instance() as any;
		instance.context.renderPreview = false;

		instance.defaultView()
		expect(spy).toHaveBeenCalled();
	});

});
