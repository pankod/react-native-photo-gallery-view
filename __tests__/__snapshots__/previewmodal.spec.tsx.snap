// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewModal Component should PreviewModal Component is render correctly 1`] = `
<react-native-mock>
  <react-native-mock>
    <react-native-mock>
      <react-native-mock />
    </react-native-mock>
  </react-native-mock>
</react-native-mock>
`;
